app/lib/features/chat_ng/widgets/reactions/reaction_detail_sheet.dart

```diff
@@ -0,0 +1,221 @@
+import 'package:acter/features/chat_ng/providers/chat_room_messages_provider.dart';
+import 'package:acter_avatar/acter_avatar.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+
+class ReactionDetailsSheet extends ConsumerStatefulWidget {
+  final String roomId;
+  final List<ReactionItem> reactions;
+
+  const ReactionDetailsSheet({
+    super.key,
+    required this.roomId,
+    required this.reactions,
+  });
+
+  @override
+  ConsumerState<ReactionDetailsSheet> createState() =>
+      _ReactionDetailsSheetState();
+}
+
+class _ReactionDetailsSheetState extends ConsumerState<ReactionDetailsSheet>
+    with TickerProviderStateMixin {
+  late TabController _tabController;
+  late List<Tab> _tabs;
+
+  @override
+  void initState() {
+    super.initState();
+    _initializeTabs();
+  }
+
+  void _initializeTabs() {
+    final total = widget.reactions.fold<int>(
+      0,
+      (sum, reaction) => sum + reaction.$2.length,
+    );
+
+    _tabs = [
+      Tab(child: Chip(label: Text('All $total'))),
+      ...widget.reactions.map(
+        (reaction) => Tab(
+          child: Chip(
+            avatar: Text(reaction.$1, style: EmojiConfig.emojiTextStyle),
+            label: Text('${reaction.$2.length}'),
+          ),
+        ),
+      ),
+    ];
+
+    _tabController = TabController(length: _tabs.length, vsync: this);
+  }
+
+  @override
+  void dispose() {
+    _tabController.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      mainAxisSize: MainAxisSize.min,
+      children: [
+        _buildTabBar(),
+        _buildTabBarView(),
+      ],
+    );
+  }
+
+  Widget _buildTabBar() {
+    return TabBar(
+      isScrollable: true,
+      padding: const EdgeInsets.all(24),
+      controller: _tabController,
+      overlayColor: WidgetStateProperty.all(Colors.transparent),
+      indicator: const BoxDecoration(
+        borderRadius: BorderRadius.all(Radius.circular(12)),
+      ),
+      indicatorSize: TabBarIndicatorSize.tab,
+      dividerColor: Colors.transparent,
+      tabs: _tabs,
+    );
+  }
+
+  Widget _buildTabBarView() {
+    final reactionData = _processReactionData();
+
+    return Flexible(
+      child: Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 14),
+        child: TabBarView(
+          controller: _tabController,
+          children: [
+            _ReactionUsersList(
+              roomId: widget.roomId,
+              users: reactionData.allUsers,
+              userReactions: reactionData.userReactions,
+            ),
+            ...widget.reactions.map(
+              (reaction) => _ReactionUsersList(
+                roomId: widget.roomId,
+                users: reactionData.reactionUsers[reaction.$1] ?? [],
+                userReactions: reactionData.userReactions,
+              ),
+            ),
+          ],
+        ),
+      ),
+    );
+  }
+
+  ({
```

Comment:
comment below said ... this is a great case for splitting the function into a utility function instead and have proper tests for it to make sure it works and doesn't break in the future.

While "copying over", I think we _do_ want to improve the code quality and in particular the code test coverage in chat NG as well.